docs/extensions.md

```diff
@@ -0,0 +1,365 @@
+    flags:
+      - name: name
+        shorthand: n
+        type: string
+        default: friend
+        usage: your name
+```
+
+## CLI API
+If the extension is making use of the `api` mode (see [Plugin modes](#plugin-modes) above), it can communicate with the CLI host via an HTTP API.
+
+### Authentication
+Upon invocation, the extension will be provided with a URI for connecting to the host CLI as well as a session token for authentication purposes.  The authentication token should be provided in the `X-CLI-Auth` header for all requests back to the CLI host.
+
+When using `api` mode, the extension's entrypoint (see [entrypoint](#entrypoint) above) will be invoked with two additional arguments representing the host HTTP server's URI and a session token, in the format `<ENTRYPOINT> <URI> <TOKEN>`.  An example invocation:
+
+```
+node index.js localhost:23855 65f6b964-a19f-4698-b2db-4e171420439b
+```
+
+Requests that do not provide the auth token will result in 401 Unauthorized responses.
+
+### Endpoint reference
+
+* [`Command`](#command)
+* [`Out`](#out)
+* [`Info`](#info)
+* [`Warn`](#warn)
+* [`Error`](#error)
+* [`Prompt`](#prompt)
+* [`Query`](#query)
+* [`Fail`](#fail)
+
+
+#### `Command`
+
+Used to collect user context for the current command execution.  Contains the command to be run as well as any flag values supplied by the user.  This is the first call the extension should make after initializing.
+
+**URL** : `/command`
+
+**Method** : `GET`
+
+**Response**
+
+```json
+{
+    "command": "hello",
+    "flags": {
+      "name": "Shelly"
+    }
+}
+```
+
+#### `Out`
+
+Used to send output to the user's terminal.
+
+**URL** : `/out`
+
+**Method** : `POST`
+
+**Request body**
+```json
+"Hello Shelly!"
+```
+
+
+#### `Info`
+
+Used to send an informational log message to the user's terminal.
+
+**URL** : `/info`
+
+**Method** : `POST`
+
+**Request body**
+```json
+"Generating hello message..."
+```
+
+#### `Warn`
+
+Used to send a warning log message to the user's terminal.
+
+**URL** : `/warn`
+
+**Method** : `POST`
+
+**Request body**
+```json
+"Hello message generation is taking longer than usual."
+```
+
+#### `Error`
+
+Used to send an error log message to the user's terminal.
+
+**URL** : `/error`
+
+**Method** : `POST`
+
+**Request body**
+```json
+"Could not generate hello message."
+```
+
+#### `Prompt`
+
+Used to collect user input via an interactive prompt.  `options`, `default`, and `placeholder` can be provided but are not required.
+
+**URL** : `/prompt`
+
+**Method** : `POST`
+
+**Request body**
+```json
+{
+  "prompt": "Who would you like to say hello to?",
+  "options": ["Shelly", "Tim", "Ramses"],
+  "default": "Shelly",
+  "placeholder": "Shelly",
+}
+```
+
+**Response**
+
+```json
+"Shelly"
+```
+
+#### `Query`
+
+Used to execute an authenticated request against the NerdGraph API.  The host CLI will use the current user's CLI profile to determine how to authenticate to the NerdGraph backend.
+
+**URL** : `/query`
```

Comment:
Should we consider `/graphql` as the endpoint here? Mutations could also be used for this endpoint and the `/graphql` endpoint seems to be a common practice for GraphQL APIs.

Reply:
👍 This isn't a graphql endpoint (yet), but still seems reasonable.